src/rpc.rs

```diff
@@ -1,103 +1,570 @@
+		let runtime_version: RuntimeVersionResult = client
+			.rpc()
+			.request("state_getRuntimeVersion", RpcParams::new())
+			.await
+			.context("Failed to retrieve runtime version")?;
+
+		let version = format!(
+			"v{}/{}/{}",
+			system_version, runtime_version.spec_name, runtime_version.spec_version,
+		);
+
+		if !expected_version.matches(&system_version, &runtime_version.spec_name) {
+			return Err(anyhow!(
+				"expected version {expected_version}, found {version}"
+			));
+		}
+		let genesis_hash = client.genesis_hash();
+		match expected_genesis_hash {
+			Some(hash) if hash != genesis_hash => {
+				return Err(anyhow!(
+					"expected genesis hash {hash}, found {genesis_hash}"
+				));
+			},
+			_ => (),
+		}
+
+		info!("Connection established to the node: {full_node_ws} <{version}>");
+		let node = Node {
+			host: full_node_ws.to_owned(),
+			system_version,
+			spec_version: client.runtime_version().spec_version,
+			genesis_hash,
+		};
+
+		Ok((client, node))
+	}
+
+	async fn connect_to_available_rpc_and_return_call<T, F, Fut>(
+		full_nodes: &[String],
+		expected_version: &ExpectedVersion<'static>,
+		expected_genesis_hash: Option<H256>,
+		mut f: F,
+	) -> Result<(avail::Client, Node, T)>
+	where
+		F: FnMut(avail::Client) -> Fut + Copy,
+		Fut: std::future::Future<Output = Result<T>>,
+	{
+		full_nodes
+			.iter()
+			.map(|address| {
+				Self::connect(address, expected_version, expected_genesis_hash)
+					.and_then(move |(client, node)| {
+						f(client.clone()).map_ok(|ret| (client, node, ret))
+					})
+					.inspect_err(move |error| warn!(address, %error, "Skipping connection"))
+			})
+			.collect::<futures::stream::FuturesUnordered<_>>()
+			.skip_while(|res| futures::future::ready(res.is_err()))
+			.map(Result::unwrap)
+			.next()
+			.await
+			.context("Failed to connect to a working node")
+	}
+
+	/// Shuffles full nodes to randomize access,
+	/// and pushes last full node to the end of a list
+	/// so we can try it if connection to other node fails
+	fn shuffle_full_nodes(full_nodes: &mut Vec<String>, last_full_node: Option<&String>) {
+		let old_len = full_nodes.len();
+		full_nodes.retain(|node| Some(node) != last_full_node);
+		full_nodes.shuffle(&mut thread_rng());
+
+		// Pushing last full node to the end of a list, if it's only one left to try
+		if let (Some(node), true) = (last_full_node, old_len != full_nodes.len()) {
+			full_nodes.push(node.clone());
+		}
+	}
+
+	pub async fn new(
+		nodes: Vec<String>,
+		expected_version: ExpectedVersion<'static>,
+		db: Option<Arc<DB>>,
+		backoff: backoff::ExponentialBackoff,
+	) -> Result<Self> {
+		let expected_genesis_hash = if let Some(db) = &db {
+			crate::data::get_genesis_hash(db.clone())?
+		} else {
+			None
+		};
+		let (client, node, ()) = Self::connect_to_available_rpc_and_return_call(
+			&nodes,
+			&expected_version,
+			expected_genesis_hash,
+			|_| futures::future::ok(()),
+		)
+		.await?;
+
+		info!(?node.genesis_hash);
```

Comment:
Can we add some context message here?

Reply:
It was like `info!("Genesis hash: {genesis_hash}")` before, so I kept it like this. What should it be?